src/shims/env.rs

```diff
@@ -91,19 +176,59 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                     .deallocate(var, None, MiriMemoryKind::Machine.into())?;
             }
             this.update_environ()?;
-            Ok(0)
+            Ok(0) // return zero on success
         } else {
             Ok(-1)
         }
     }
 
+    fn setenvironmentvariablew(
+        &mut self,
+        name_op: OpTy<'tcx, Tag>, // LPCWSTR lpName,
+        value_op: OpTy<'tcx, Tag>, // LPCWSTR lpValue,
+    ) -> InterpResult<'tcx, i32> {
+        let mut this = self.eval_context_mut();
+        this.assert_target_os("windows", "SetEnvironmentVariableW");
+
+        let name_ptr = this.read_scalar(name_op)?.not_undef()?;
+        let value_ptr = this.read_scalar(value_op)?.not_undef()?;
+
+        let mut new = None;
+        if !this.is_null(name_ptr)? {
```

Comment:
Is the name really allowed to be NULL? (For `setenv`, this is explicitly allowed in the docs, and specified to return EINVAL. But I see nothing like that in the Windows docs so I assume it's not allowed, and should cause UB.)

Reply:
Also in case of an error the docs say to inspect `GetLastError`, so we should set the last error (like we do above for getting an env var). But it doesn't say which error code is even possible, and...

> The Return Value section of the documentation for each function that sets the last-error code notes the conditions under which the function sets the last-error code.

... but for SetEnvironmentVariableW it says nothing. So I think it's fine to raise UB if the ptr is NULL, and to do `throw_unsup_format!` if the name is empty or contains a `=`.